### PR TITLE
Limit entity push abuse

### LINKED ISSUE #1278: Entity push speed/strafe in 1.8 server
### Describe the bypass and how to replicate it

Place many boats on ground and use strafe on them, you will get less flags

1.11+ vanilla clients in 1.8 server do not appear to be subject to collisions from other entities

Can we disable entity push handler when the maxEntityCramming gamerule = 0? Players won't get push by other player in this context.

### Grim version

2.3.48

### Server version

Paper 1.8.8

### Plugins

eu.loyisa.cn

### LINKED ISSUE #1414: Speed Bypass
### Describe the bypass and how to replicate it

package net.ccbluex.liquidbounce.features.module.modules.movement

import net.ccbluex.liquidbounce.event.*
import net.ccbluex.liquidbounce.features.module.Module
import net.ccbluex.liquidbounce.features.module.ModuleCategory
import net.ccbluex.liquidbounce.features.module.ModuleInfo
import net.ccbluex.liquidbounce.injection.backend.unwrap
import net.ccbluex.liquidbounce.utils.extensions.getDistanceToEntityBox
import net.ccbluex.liquidbounce.value.BoolValue
import net.ccbluex.liquidbounce.value.FloatValue
import net.ccbluex.liquidbounce.value.IntegerValue
import net.minecraft.entity.EntityLivingBase
import net.minecraft.network.play.server.SPacketPlayerPosLook

@ModuleInfo(name = "EntitySpeed", description = "EntitySpeed", category = ModuleCategory.MOVEMENT)
class EntitySpeed : Module() {
    private val onlyAir = BoolValue("OnlyAir",false)
    private val okstrafe = BoolValue("Strafe",false)
    private val keepSprint = BoolValue("KeepSprint",false)
    private val speedUp = BoolValue("SpeedUp",false)
    private val hurtcancle = BoolValue("Hurtcancle",false)

    private val speedplus = BoolValue("Speedplus",true)
    private val speedplusvaule = IntegerValue("Speedplusvaule", 0, 0, 10)
    private val speedd = IntegerValue("Speed", 0, 0, 10)
    private var speed = 0
    private val distance = FloatValue("speedRange", 0f, 0f, 3f)
    private val distance2 = FloatValue("starfeRange", 0f, 0f, 3f)
    private var speeded = false
    private var f = 0
    var strgo = false
    private var pre = false
    var sprint = false

    override fun onEnable() {
        speeded = false
    }
    @EventTarget
    fun onPacket(event: PacketEvent) {
        val packet = event.packet.unwrap()
        if (packet is SPacketPlayerPosLook){
            f = 10
        }
    }
    @EventTarget
    fun onUpdate(event: UpdateEvent) {
        speed = speedd.get()
        f--
        if (f >= 0) {
            mc.thePlayer!!.motionX *= 1
            mc.gameSettings.keyBindForward.pressed = false
            mc.gameSettings.keyBindLeft.pressed = false
            mc.gameSettings.keyBindBack.pressed = false
            mc.gameSettings.keyBindRight.pressed = false
            mc.gameSettings.keyBindJump.pressed = false
            return
        }
        if (mc.thePlayer!!.hurtTime > 0 && hurtcancle.get()) return
        for (entity in mc.theWorld!!.loadedEntityList) {
            if (entity.unwrap() is EntityLivingBase && entity.entityId != mc.thePlayer!!.entityId && mc.thePlayer!!.getDistanceToEntityBox(
                    entity
                ) <= distance.get() && (!onlyAir.get() || !mc.thePlayer!!.onGround)
            ) {
                if(speedplus.get()){
                    speed = (speedd.get()+speedplusvaule.get()*(1- mc.thePlayer!!.getDistanceToEntityBox(entity)/distance.get())).toInt()
                }
                if (speedUp.get()) {
                    mc.thePlayer!!.motionX *= (1 + (speed * 0.01))
                    mc.thePlayer!!.motionZ *= (1 + (speed * 0.01))
                }
                if (keepSprint.get()) {
                    sprint = true
                }
                if (okstrafe.get() && mc.thePlayer!!.getDistanceToEntityBox(
                        entity
                    ) <= distance2.get()) {
                    strgo = true
                }
                return
            }
            sprint = false
            strgo = false
        }
    }

}

### Grim version

grimac-2.3.61.jar

### Server version

1.8.8

### Plugins

ViaVersion-4.9.2.jar grimac-2.3.61.jar

### LINKED ISSUE #1511: Speed bypass
### Describe the bypass and how to replicate it

if ((mode.is(Mode.GrimEntity2)|| mode.is(Mode.GrimCombo)) && MovementUtility.isMoving()) {
            int collisions = 0;
            for (Entity ent : mc.world.getEntities())
                if (ent != mc.player && (ent instanceof LivingEntity || ent instanceof BoatEntity) && mc.player.getBoundingBox().expand(1.0).intersects(ent.getBoundingBox()))
                    collisions++;

            double[] motion = MovementUtility.forward(0.08 * collisions);
            mc.player.addVelocity(motion[0], 0.0, motion[1]);
        }

### Grim version

2.3.64

### Server version

1.17.1

### Plugins

GrimAC PacketEvents

### LINKED ISSUE #1615: Speed bypasss
### Describe the bypass and how to replicate it

1. Donwload LuqiudBounce
2. Set parametrs how on video
3.  Done

### Grim version

Latest Grim 2.3.66

### Server version

Purpur 1.20.6
And video is here -  https://youtu.be/O6xcFzt52Kg 
On video i forgot to add parametrs off function LuqiudBounce, its on screenshoot
And code of fix (Its in MovementTicker) ---> handleplayercolissions

 ITS ONLY HOTFIX, AND I DON'T ANSWERING ABOUT ANY FALSES IN THIS FIX!!! And etc. 

public static void handleEntityCollisions(GrimPlayer player) {
        // 1.7 and 1.8 do not have player collision
        if (player.getClientVersion().isOlderThan(ClientVersion.V_1_9)) return;

        int possibleCollidingEntities = 0;

        // Players in vehicles do not have collisions
        if (!player.compensatedEntities.getSelf().inVehicle()) {
            // Calculate the offset of the player to colliding other stuff
            SimpleCollisionBox playerBox = GetBoundingBox.getBoundingBoxFromPosAndSize(player, player.lastX, player.lastY, player.lastZ, 0.6f, 1.8f);
            SimpleCollisionBox expandedPlayerBox = playerBox.copy().expandToAbsoluteCoordinates(player.x, player.y, player.z).expand(1);

            for (PacketEntity entity : player.compensatedEntities.entityMap.values()) {
                if (!entity.isPushable())
                    continue;

                SimpleCollisionBox entityBox = entity.getPossibleCollisionBoxes();

                if (expandedPlayerBox.isCollided(entityBox))
                    possibleCollidingEntities++;
            }
        }
// fix hacking with colliding with entity
        double speedLimit = 0.42 + (possibleCollidingEntities * 0.01);

        if (player.isSneaking) {
            if (player.lastOnGround) {
                if (player.clientVelocity.lengthSquared() > speedLimit * speedLimit) {
                    player.clientVelocity = player.clientVelocity.normalize().multiply(speedLimit);
                }
            }
            if (player.isFlying && player.wasFlying) {
                {
                    return;
                }
            } else if (player.wasSneaking) {
                if (player.clientVelocity.lengthSquared() > speedLimit * speedLimit) {
                    player.clientVelocity = player.clientVelocity.normalize().multiply(speedLimit);
                }
            }

            if (player.isSprinting) {
                if (player.lastOnGround) {
                    if (player.clientVelocity.lengthSquared() > speedLimit * speedLimit) {
                        player.clientVelocity = player.clientVelocity.normalize().multiply(speedLimit);
                    }
                }
                if (player.isFlying && player.wasFlying) {
                    {
                        return;
                    }

                }
            } else if (player.lastSprinting) {
                if (player.clientVelocity.lengthSquared() > speedLimit * speedLimit) {
                    player.clientVelocity = player.clientVelocity.normalize().multiply(speedLimit);
                }
            }

            if (player.isGliding && possibleCollidingEntities > 0) {
                // Horizontal starting movement affects vertical movement with elytra, hack around this.
                // This can likely be reduced but whatever, I don't see this as too much of a problem
                player.uncertaintyHandler.yNegativeUncertainty -= 0.05;
                player.uncertaintyHandler.yPositiveUncertainty += 0.05;
            }

            player.uncertaintyHandler.collidingEntities.add(possibleCollidingEntities);
        }
    }
    
    ITS ONLY HOTFIX, AND I DON'T ANSWERING ABOUT ANY FALSES IN THIS FIX!!! And etc. 

### Plugins

[23:26:16 INFO]: Paper Plugins: (2):
[23:26:16 INFO]:  - BKCommonLib, FancyHolograms
[23:26:16 INFO]: Bukkit Plugins: (45):
[23:26:16 INFO]:  - AdvancedCrates, ajLeaderboards, AntiCheatAddition, BetterRTP, BetterStructures, Boss, Chatty, Chunky, Citizens, CMI
[23:26:16 INFO]:  CMILib, ConditionalCommands, ConditionalEvents, DecentHolograms, DeluxeMenus, ExecutableItems, FacilisCommon, GrimAC, Iris, LiteBans
[23:26:16 INFO]:  LPX, LuckPerms, Mars, MarsGenerator, MoonGenerator, MultiWorld, My_Worlds, packetevents, Pl-Hide-Pro, PlaceholderAPI
[23:26:16 INFO]:  PlasmoVoice, PlayerPoints, PlugManX, ProtectionStones, ProtocolLib, SCore, SpacePlugin, spark, TAB, Vault
[23:26:16 INFO]:  ViaBackwards, ViaVersion, W
Снимок экрана_20240724_224827
orldEdit, WorldGuard, WorldGuardExtraFlags

### PR DESCRIPTION
**Pull request by SamB440**
I tried to add full simulation of entity pushing but there were annoying edge-cases.

This checks for player and entity teams to see if the entity can truly be pushed. On 1.8 servers if prevent-collision is enabled in Via config, 1.9+ players can never be pushed by an entity. Additionally the bounding box calculation has been changed to reduce abuse.

Fixes #1511 
Fixes #1414
Fixes #1278
Fixes #1615

### CODE REVIEW

**Comment by MatrixU5er** [comment]
A good patch that fixed the speed of all the clients I have

File: src/main/java/ac/grim/grimac/utils/team/EntityTeam.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        return name;
+    }
+
+    public Set<String> getEntries() {
+        return entries;
+    }
+
+    public WrapperPlayServerTeams.CollisionRule getCollisionRule() {
+        return collisionRule;
+    }
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/utils/team/EntityTeam.java]: `@Getter`?

Comment (SamB440) [reply src/main/java/ac/grim/grimac/utils/team/EntityTeam.java]: I like to avoid Lombok

Comment (ManInMyVan) [reply src/main/java/ac/grim/grimac/utils/team/EntityTeam.java]: What's wrong with lombok?

**Comment by delyfssflaggedagain** [comment]
u can fix this by 2 lines of code lol

**Comment by Souipi** [comment]
> u can fix this by 2 lines of code lol

Do it so ?

**Comment by SamB440** [comment]
> u can fix this by 2 lines of code lol

Show me proof that you can get any reasonable speed advantage with these changes.

File: src/main/java/ac/grim/grimac/utils/team/EntityTeam.java (hunk 1 above)
Comment (HackProtectorX) [reply src/main/java/ac/grim/grimac/utils/team/EntityTeam.java]: lombok bad

**Pull request merged**